src/pip/_internal/operations/prepare.py

```diff
@@ -448,12 +459,48 @@ def _get_linked_req_hashes(self, req):
         # showing the user what the hash should be.
         return req.hashes(trust_internet=False) or MissingHashes()
 
+    def _fetch_metadata(preparer, link):
```

Comment:
This is probably something we can do in a follow-up, but we should add logging here, to denote which branch we're going through, and introduce early returns to make the flow cleaner.

```py
if not preparer.use_lazy_wheel:
    return None

if preparer.require_hashes:
    logger.debug(...)
    return None

if link.is_file or not link.is_wheel:
    logger.debug(...)
    return None

<all the current logic>
```

Reply:
Let's do this in a follow up. @McSinyx wanna pick this up? :)

Reply:
Yes I'm glad to.